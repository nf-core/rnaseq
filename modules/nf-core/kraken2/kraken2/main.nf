process KRAKEN2_KRAKEN2 {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mulled-v2-8706a1dd73c6cc426e12dd4dd33a5e917b3989ae:c8cbdc8ff4101e6745f8ede6eb5261ef98bdaff4-0' :
        'biocontainers/mulled-v2-8706a1dd73c6cc426e12dd4dd33a5e917b3989ae:c8cbdc8ff4101e6745f8ede6eb5261ef98bdaff4-0' }"

    input:
    tuple val(meta), path(reads)
    path  db
    val save_output_fastqs
    val save_reads_assignment

    output:
    tuple val(meta), path('*.classified{.,_}*')     , optional:true, emit: classified_reads_fastq
    tuple val(meta), path('*.unclassified{.,_}*')   , optional:true, emit: unclassified_reads_fastq
    tuple val(meta), path('*classifiedreads.txt')   , optional:true, emit: classified_reads_assignment
    tuple val(meta), path('*report.txt')                           , emit: report
    path "versions.yml"                                            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def paired       = meta.single_end ? "" : "--paired"
    def classified   = meta.single_end ? "${prefix}.classified.fastq"   : "${prefix}.classified#.fastq"
    def unclassified = meta.single_end ? "${prefix}.unclassified.fastq" : "${prefix}.unclassified#.fastq"
    def classified_option = save_output_fastqs ? "--classified-out ${classified}" : ""
    def unclassified_option = save_output_fastqs ? "--unclassified-out ${unclassified}" : ""
    def readclassification_option = save_reads_assignment ? "--output ${prefix}.kraken2.classifiedreads.txt" : "--output /dev/null"
    def compress_reads_command = save_output_fastqs ? "pigz -p $task.cpus *.fastq" : ""

    """
    kraken2 \\
        --db $db \\
        --threads $task.cpus \\
        --report ${prefix}.kraken2.report.txt \\
        --gzip-compressed \\
        $unclassified_option \\
        $classified_option \\
        $readclassification_option \\
        $paired \\
        $args \\
        $reads

    $compress_reads_command

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kraken2: \$(echo \$(kraken2 --version 2>&1) | sed 's/^.*Kraken version //; s/ .*\$//')
        pigz: \$( pigz --version 2>&1 | sed 's/pigz //g' )
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def paired       = meta.single_end ? "" : "--paired"
    def classified   = meta.single_end ? "${prefix}.classified.fastq.gz"   : "${prefix}.classified_1.fastq.gz ${prefix}.classified_2.fastq.gz"
    def unclassified = meta.single_end ? "${prefix}.unclassified.fastq.gz" : "${prefix}.unclassified_1.fastq.gz ${prefix}.unclassified_2.fastq.gz"
    def readclassification_option = save_reads_assignment ? "--output ${prefix}.kraken2.classifiedreads.txt" : "--output /dev/null"
    def compress_reads_command = save_output_fastqs ? "pigz -p $task.cpus *.fastq" : ""

    """
    touch ${prefix}.kraken2.report.txt
    if [ "$save_output_fastqs" == "true" ]; then
        touch $classified
        touch $unclassified
    fi
    if [ "$save_reads_assignment" == "true" ]; then
        touch ${prefix}.kraken2.classifiedreads.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kraken2: \$(echo \$(kraken2 --version 2>&1) | sed 's/^.*Kraken version //; s/ .*\$//')
        pigz: \$( pigz --version 2>&1 | sed 's/pigz //g' )
    END_VERSIONS
    """

}
