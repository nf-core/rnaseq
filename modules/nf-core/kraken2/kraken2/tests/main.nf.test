nextflow_process {
    name "Test Process KRAKEN2_KRAKEN2"
    script "../main.nf"
    process "KRAKEN2_KRAKEN2"
    tag "modules"
    tag "modules_nfcore"
    tag "untar"
    tag "kraken2"
    tag "kraken2/kraken2"

    setup {
        run("UNTAR") {
            script "modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [],
                    file(
                        params.modules_testdata_base_path + "genomics/sarscov2/genome/db/kraken2.tar.gz",
                        checkIfExists: true
                    )
                ])
                """
            }
        }
    }

    test("sarscov2 illumina single end [fastq]") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    [ file(
                        params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz",
                        checkIfExists: true
                    )]
                ]
                input[1] = UNTAR.out.untar.map{ it[1] }
                input[2] = true
                input[3] = false
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    process.out.report,
                    process.out.versions,
                ).match()
            },
            { assert process.out.classified_reads_fastq.get(0).get(1) ==~ ".*/test.classified.fastq.gz" },
            { assert process.out.unclassified_reads_fastq.get(0).get(1) ==~ ".*/test.unclassified.fastq.gz" },
            )
        }
    }

    test("sarscov2 illumina paired end [fastq]") {
        when {
            params {
                outdir   = "$outputDir"
            }

            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(
                            params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz",
                            checkIfExists: true
                        ),
                        file(
                            params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_2.fastq.gz",
                            checkIfExists: true
                        )

                        ]
                    ]
                input[1] = UNTAR.out.untar.map{ it[1] }
                input[2] = true
                input[3] = false
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    process.out.report,
                    process.out.versions,
                ).match()
            },
            { assert process.out.classified_reads_fastq.get(0).get(1).get(0)
                ==~ ".*/test.classified_1.fastq.gz" },
            { assert process.out.classified_reads_fastq.get(0).get(1).get(1)
                ==~ ".*/test.classified_2.fastq.gz" },
            { assert process.out.unclassified_reads_fastq.get(0).get(1).get(0)
                ==~ ".*/test.unclassified_1.fastq.gz" },
            { assert process.out.unclassified_reads_fastq.get(0).get(1).get(1)
                ==~ ".*/test.unclassified_2.fastq.gz" },
            )
        }
    }

    test("sarscov2 illumina single end [fastq] + save_reads_assignment") {
        when {
            params {
                outdir   = "$outputDir"
            }

            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    [ file(
                        params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz",
                        checkIfExists: true
                    )]
                ]
                input[1] = UNTAR.out.untar.map{ it[1] }
                input[2] = false
                input[3] = true
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(
                    process.out.report,
                    process.out.classified_reads_assignment,
                    process.out.versions,
                ).match()
            },
            )
        }
    }
}
