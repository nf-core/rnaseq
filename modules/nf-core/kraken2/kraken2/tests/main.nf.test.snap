{
    "sarscov2 illumina single end [fastq]": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.kraken2.report.txt:md5,4227755fe40478b8d7dc8634b489761e"
                ]
            ],
            [
                "versions.yml:md5,79adf2ca1cfc625cb77e391b27142c43"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-04T18:47:03.745692"
    },
    "sarscov2 illumina paired end [fastq]": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.kraken2.report.txt:md5,4227755fe40478b8d7dc8634b489761e"
                ]
            ],
            [
                "versions.yml:md5,79adf2ca1cfc625cb77e391b27142c43"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-04T18:47:13.75649"
    },
    "sarscov2 illumina single end [fastq] + save_reads_assignment": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.kraken2.report.txt:md5,4227755fe40478b8d7dc8634b489761e"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.kraken2.classifiedreads.txt:md5,e7a90531f0d8d777316515c36fe4cae0"
                ]
            ],
            [
                "versions.yml:md5,79adf2ca1cfc625cb77e391b27142c43"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-04T18:47:22.459465"
    }
}